### PR TITLE
Fix CUB `min`/`max` initial values

### LINKED ISSUE #8180: `cupy.min()` mishandles infinity
I see the same problem with cupy version 13.0.0:
```python
In [3]: cp.min(cp.array([cp.inf, cp.inf]))
Out[3]: array(1.79769313e+308)
```

Is this because the bug has come back, is caused by something else or did the fix never make it in to the release?

Environment info:
```
OS                           : Linux-5.15.0-94-generic-x86_64-with-glibc2.35
Python Version               : 3.12.1
CuPy Version                 : 13.0.0
CuPy Platform                : NVIDIA CUDA
NumPy Version                : 1.26.3
SciPy Version                : 1.12.0
Cython Build Version         : 0.29.37
Cython Runtime Version       : 3.0.8
CUDA Root                    : /usr/local/cuda
nvcc PATH                    : /usr/local/cuda/bin/nvcc
CUDA Build Version           : 12000
CUDA Driver Version          : 12020
CUDA Runtime Version         : 12000 (linked to CuPy) / 12010 (locally installed)
cuBLAS Version               : (available)
cuFFT Version                : 11002
cuRAND Version               : 10304
cuSOLVER Version             : (11, 4, 4)
cuSPARSE Version             : (available)
NVRTC Version                : (12, 3)
Thrust Version               : 200001
CUB Build Version            : 200200
Jitify Build Version         : c08b8c6
cuDNN Build Version          : None
cuDNN Version                : None
NCCL Build Version           : None
NCCL Runtime Version         : None
cuTENSOR Version             : None
cuSPARSELt Build Version     : None
Device 0 Name                : Tesla T4
Device 0 Compute Capability  : 75
Device 0 PCI Bus ID          : 0000:3B:00.0
Device 1 Name                : Tesla T4
Device 1 Compute Capability  : 75
Device 1 PCI Bus ID          : 0000:5E:00.0
Device 2 Name                : Tesla T4
Device 2 Compute Capability  : 75
Device 2 PCI Bus ID          : 0000:AF:00.0
Device 3 Name                : Tesla T4
Device 3 Compute Capability  : 75
Device 3 PCI Bus ID          : 0000:D8:00.0
```

_Originally posted by @betatim in https://github.com/cupy/cupy/issues/7424#issuecomment-1941010817_

### PR DESCRIPTION
**Pull request by leofang**
Close #8180. 

The current implementation requires C++17 in order to avoid unnecessary template instantiations. Not sure if there's any better approach without bumping the C++ version, or bumping does not matter much... 

Things to check:
- binary size
- glibc version

### CODE REVIEW

File: install/cupy_builder/_compiler.py [hunk 1]
```diff
@@ -215,14 +198,14 @@ def _compile_unix_nvcc(self, obj: str, src: str, ext: Extension) -> None:
...
-            postargs += [f'-t{num_threads}']
-        else:
-            postargs += ['--std=c++11']
-        postargs += ['-Xcompiler=-fno-gnu-unique']
+            '-Xfatbin=-compress-all', '-O2', '--compiler-options="-fPIC"',
+            '--expt-relaxed-constexpr']
+        num_threads = int(os.environ.get('CUPY_NUM_NVCC_THREADS', '2'))
+        # Note: we only support CUDA 11.2+ since CuPy v13.0.0.
+        # Bumping C++ standard from C++14 to C++17 for "if constexpr"
+        postargs += ['--std=c++17',
```
Comment (leofang) [inline install/cupy_builder/_compiler.py:206]: @kmaehashi any concern?

Comment (kmaehashi) [reply install/cupy_builder/_compiler.py:206]: No problem, for wheel build we use CentOS 7 with gcc 7.3.1 which supports C++17.

**Comment by leofang** [comment]
I see no size in crease in the binary.

**Comment by leofang** [comment]
@miscco @kmaehashi it seems the Windows CUDA 11.2 CI keeps failing at build time when compiling `cupy/cuda/cupy_thrust.cu`. Any idea what could be the problem? I can't tell if it is an MSVC issue or libcudacxx issue...

**Comment by leofang** [comment]
(Bumping priority to high as @betatim told me offline this is a blocker to scikit-learn)

**Comment by miscco** [comment]
I am reading errors such as:
```
00:01:34.583039 STDERR 1584] | a.cpp |  
-- | -- | --
00:01:34.661165 STDERR 1584] | C:\Windows\TEMP\flexci\run-00324733\tmp\tmpky6kni1b\a.cpp(1): fatal error C1083: Cannot open include file: 'nccl.h': No such file or directory |  
00:01:34.679567 STDERR 1584] | command 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2017\\BuildTools\\VC\\Tools\\MSVC\\14.16.27023\\bin\\HostX86\\x64\\cl.exe' failed with exit code 2
```

That suggests to me that this is not a cccl issue

**Comment by leofang** [comment]
Sorry, CuPy's build system can emit confusing logs, things happen after the module detection section
```
[00:01:38.386631 STDERR 1584]](https://ci.preferred.jp/cupy.win.cuda112/158267/#L328)	  Modules:	
[00:01:38.387607 STDERR 1584]](https://ci.preferred.jp/cupy.win.cuda112/158267/#L329)	    cuda      : Yes (version 11020)	
[00:01:38.387607 STDERR 1584]](https://ci.preferred.jp/cupy.win.cuda112/158267/#L330)	    cusolver  : Yes	
[00:01:38.387607 STDERR 1584]](https://ci.preferred.jp/cupy.win.cuda112/158267/#L331)	    cudnn     : Yes (version 8801)
...
```
are where the real compilations start. So, scrolling down from there, the first error is
```
[00:05:51.739833 STDERR 1584]](https://ci.preferred.jp/cupy.win.cuda112/158267/#L6077)	  C:/Windows/TEMP/flexci/run-00324733/work/src/cupy/_core/include/cupy/_cccl/cub\cub/util_cpp_dialect.cuh:129: warning: CUB requires at least MSVC 2019 (19.20/16.0/14.20). MSVC 2017 is deprecated but still supported. MSVC 2017 support will be removed in a future release. Define CUB_IGNORE_DEPRECATED_CPP_DIALECT to suppress this message.	
[00:05:51.739833 STDERR 1584]](https://ci.preferred.jp/cupy.win.cuda112/158267/#L6080)	  c:\windows\temp\flexci\run-00324733\work\src\cupy\_core\include\cupy\_cccl\libcudacxx\cuda\std\detail\libcxx\include\__concepts\../__concepts/convertible_to.h(41): error: this declaration has no storage class or type specifier	
[00:05:51.739833 STDERR 1584]](https://ci.preferred.jp/cupy.win.cuda112/158267/#L6082)		
[00:05:51.739833 STDERR 1584]](https://ci.preferred.jp/cupy.win.cuda112/158267/#L6083)	  c:\windows\temp\flexci\run-00324733\work\src\cupy\_core\include\cupy\_cccl\libcudacxx\cuda\std\detail\libcxx\include\__concepts\../__concepts/convertible_to.h(41): error: a value of type "const char *" cannot be used to initialize an entity of type "int"	
[00:05:51.739833 STDERR 1584]](https://ci.preferred.jp/cupy.win.cuda112/158267/#L6086)		
[00:05:51.739833 STDERR 1584]](https://ci.preferred.jp/cupy.win.cuda112/158267/#L6087)	  c:\windows\temp\flexci\run-00324733\work\src\cupy\_core\include\cupy\_cccl\libcudacxx\cuda\std\detail\libcxx\include\__concepts\../__concepts/convertible_to.h(45): error: expected a ";"	1.2s
[00:05:51.739833 STDERR 1584]](https://ci.preferred.jp/cupy.win.cuda112/158267/#L6089)
```
Note that we use the same CCCL submodule for all CUDA & OS versions, but only this particular CI fails at build time.

**Comment by miscco** [comment]
We tracked this down in an internal conversation. This is a bug in cccl in the respective commit used to test.

There is either the possibility to put a minimal fix on top of the cccl checkout or move to 2.3.x altogether

**Comment by leofang** [comment]
> There is either the possibility to put a minimal fix on top of the cccl checkout or move to 2.3.x altogether

Yes, let me verify whether the patch works first, and then we can decide the fix strategy (update CCCL now or later).

**Comment by leofang** [comment]
OK, the win.cuda112 CI has run over 10 mins and is still running, so hopefully we've moved past the build phase, meaning @miscco's patch is working 😄

Here's what I am thinking @kmaehashi, let me know your thoughts:
- In this PR, we use Michael's patch (currently hosted in my fork) to fix the original min/max bug, but not update the CCCL submodule. We then backport this PR to v13.1.0 to unblock scikit-learn. 
- Then, in Michael's other PR (#8221), we update the CCCL submodule to v2.3.1 or later. We decide by then if it needs back porting. (In fact, Michael's PR was meant to assist CCCL update.)

**Comment by leofang** [comment]
CI is all green!

**Comment by miscco** [comment]
💚

File: .gitmodules [hunk 2]
```diff
@@ -1,6 +1,6 @@
 [submodule "third_party/cccl"]
 	path = third_party/cccl
-	url = https://github.com/NVIDIA/cccl.git
+	url = https://github.com/leofang/cccl.git
```
Comment (kmaehashi) [inline .gitmodules:3]: I've made a fork under CuPy org and mirrored your branch at https://github.com/cupy/cccl/tree/fix_cupy_8203!

```suggestion
	url = https://github.com/cupy/cccl.git
```

**Review by kmaehashi — CHANGES_REQUESTED** [review]
LGTM! Could you double check @asi1024?

File: .gitmodules (hunk 2 above)
Comment (leofang) [reply .gitmodules:3]: Great, thanks @kmaehashi! Could you grant me (and @miscco perhaps) write/developer access to the fork, please? I think in the near future there will be some CCCL-related work needed, so it's easier for us if we can push to the fork.

**Comment by asi1024** [comment]
LGTM excepts @kmaehashi's comment.

**Review by kmaehashi — APPROVED** [review]

**Pull request merged**